packages/angular/ssr/src/routes/route-tree.ts

```diff
@@ -0,0 +1,224 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.dev/license
+ */
+
+import { joinUrlParts, stripTrailingSlash } from '../utils/url';
+
+/**
+ * Interface representing the structure of a route node in a nested object representation.
+ */
+export interface RouteNodeObjectRepresentation {
+  /** Optional metadata associated with this route node. */
+  metadata?: RouteTreeNodeMetadata;
+
+  /** Optional children nodes of this route node, keyed by their segment values. */
+  children?: Record<string, RouteNodeObjectRepresentation>;
+}
+
+/**
+ * Metadata for a route tree node.
+ */
+export interface RouteTreeNodeMetadata {
+  /** Optional redirect route associated with this node. */
+  redirectTo?: string;
+
+  /** The route segment for this node. */
+  route: string;
+}
+
+/**
+ * A node in a route tree.
+ */
+interface RouteTreeNode {
+  /** The value associated with this node (a route segment or wildcard). */
+  segment: string;
+
+  /** The insertion index of the route stored at this node or a descendant. */
+  insertionIndex: number;
+
+  /** A map of child nodes, keyed by their values. */
+  children?: Map<string, RouteTreeNode>;
+
+  /** Indicates if this node is an end node. */
+  isEndNode?: boolean;
+
+  /** Optional metadata associated with this node. */
+  metadata?: RouteTreeNodeMetadata;
+}
+
+/**
+ * A route tree implementation for efficient route matching with wildcards.
+ */
+export class RouteTree {
+  /** The root node of the route tree. */
+  private root: RouteTreeNode;
+
+  /** A counter to track the insertion order of routes. */
+  private insertionIndexCounter = 0;
+
+  /**
+   * Creates a new RouteTree.
+   * @param rootSegment - (Optional) The segment for the root node (defaults to an empty string).
+   */
+  constructor(private rootSegment = '') {
+    this.root = {
+      segment: rootSegment,
+      insertionIndex: -1,
+    };
+  }
+
+  /**
+   * Inserts a route into the route tree.
+   * @param route - The route to insert.
+   * @param metadata - The metadata associated with the route.
+   */
+  insert(route: string, metadata: Omit<RouteTreeNodeMetadata, 'route'>): void {
+    let node = this.root;
+    const normalizedRoute = stripTrailingSlash(route);
+    const segments = normalizedRoute.split('/');
+
+    for (const segment of segments) {
+      // Parameterized route segment, replace it with a wildcard
+      const normalizedSegment = segment[0] === ':' ? '*' : segment;
+      let childNode = node.children?.get(normalizedSegment);
+
+      if (!childNode) {
+        childNode = {
+          segment: normalizedSegment,
+          insertionIndex: -1,
+        };
+        node.children ??= new Map();
```

Comment:
Originally, my goal was to avoid creating potentially hundreds of redundant `Map` instances in the leaf nodes for large applications, especially when considering i18n.

Reply:
I personally wouldn't worry too much about optimizing at this stage until we have a sense of where the performance bottlenecks are, but I'll defer to you on which improvements are likely to be significant enough wins to be worth landing up front.